### PR TITLE
added prototype retetion time indexing, solves #33

### PR DESCRIPTION
**Pull request by michaelwitting**
Added first version of retention time indexing (issue #33 ). Currently linear interpolation is the only way of performing indexing.

### CODE REVIEW

File: R/retentionIndexing.R [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+#' rti <- data.frame(rtime = c(1,2,3),
+#' rindex = c(100,200,300))
+#' rtime <- c(1.5, 2.5)
+#' rtimeIndexing(rtime, rti)
+#' 
+rtimeIndexing <- function(x,
+                          y,
+                          FUN = rtiLinear) {
+  
+  unlist(lapply(x, FUN = FUN, rti = y))
```
Comment (sgibb) [inline R/retentionIndexing.R:34]: The `approx` function is vectorized. So you could just use:
```r
FUN <- match.fun(FUN)
do.call(FUN, list(x = x, rti = y))
# or even FUN(x = x, y = rti)
```
If you think you really need a kind of *apply function here I would suggest `vapply`.

BTW: I would allow additional arguments to `FUN` (`...`).

File: R/retentionIndexing.R [hunk 2]
```diff
@@ -0,0 +1,48 @@
+#' @title Convert retention times to retention indices
```
Comment (sgibb) [inline R/retentionIndexing.R:1]: Is that similar to [`MSnbase::alignRt()`](https://github.com/lgatto/MSnbase/blob/655ff179bf149e5abc45fccf4f06a6a6fd34e32a/R/functions-Chromatogram.R#L149-L230)?

I don't know whether this drift is always linear but if not it may be worth to consider some kind of non-linear alignment functions, e.g. [`synapter:::modelRetTime()`](https://github.com/lgatto/synapter/blob/24c0b7b0ae270143bedf16517b6c376a43f22545/R/utils.R#L69-L99) (lowess) or [`MALDIquant::determineWarpingFunctions()`](https://github.com/sgibb/MALDIquant/blob/master/R/determineWarpingFunctions-functions.R) (used for aligning MS1 to MS1 not rtimes but should be easily adoptable, supports lowess, linear, quadratic, cubic).

**Review by sgibb — CHANGES_REQUESTED** [review]
I am not familiar with Metabolomics. Maybe my questions are a little bit stupid.

Are the `inst/retentionIndex/*.txt` files general usable or just for demonstration?

File: R/retentionIndexing.R (hunk 1 above)
Comment (michaelwitting) [reply R/retentionIndexing.R:34]: Thank you for the suggestion. `approx` is only one of the functions I would like to implement, there are also cubic splines etc.

File: R/retentionIndexing.R (hunk 2 above)
Comment (michaelwitting) [reply R/retentionIndexing.R:1]: In this case it is not about drift normalization but converting absolute retention times (RTs) to retention indices (RIs), which makes retention data comparable across two different LC systems. Even when using the same column, solvent and gradient differences in absolute RTs are found due to differences like the pumping system, dead volume etc... The conversion from RTs to RIs normalizes for this factors, which allows to use retention data for metabolite identification. This step is often done in GC-MS using the Kovats index.

Comment (michaelwitting) [reply R/retentionIndexing.R:1]: And the files in `inst/retentionIndex` are just for demonstration. One has to first determine the RTs of the indexing substances and then can convert RTs to RIs. The specific example is from an upcoming publication of myself, in which I present an RI system for LC-MS based metabolomics.

**Review by sgibb — APPROVED** [review]
Thanks for the clarification.

File: inst/retentionIndex/metabolites.txt [hunk 3]
```diff
@@ -0,0 +1,77 @@
+name	rtime	rindex_manual
+VITAMIN D2	NA	NA
+SQUALENE	15.66	1709.876543
```
Comment (jorainer) [inline inst/retentionIndex/metabolites.txt:3]: Hm, I find it a little unintuitive to call a floating point number *index* - I thought index would be an `integer`? Wouldn't it be *better* to provide integer indices instead?

File: vignettes/MetaboCoreUtils.Rmd [hunk 4]
```diff
@@ -126,6 +126,48 @@ countElements(frml)
...
+## Retetion time indexing
+
+Retention times are often not directly comparable between two LC-MS systems, even
+if nominally the same separation method is used. Conversion of retention times to
+retetion indices can overcome this issue. The `MetaboCoreUtils` package provides
+a function to perform this conversion. Below we use an example based on indexing
+with a homologoues series af N-Alkyl-pyridinium sulfonates (NAPS).
+
+```{r}
+rti <- read.table(system.file("retentionIndex/rti.txt",
```
Comment (jorainer) [inline vignettes/MetaboCoreUtils.Rmd:138]: please use here `system.file("retentionIndex", "rti.txt", package = "MetaboCoreUtils")`. This ensures the code to work for sure on any platform

File: vignettes/MetaboCoreUtils.Rmd [hunk 5]
```diff
@@ -126,6 +126,48 @@ countElements(frml)
...
+                              package = "MetaboCoreUtils"),
+                  header = TRUE,
+                  sep = "\t")
+  
+rtime <- read.table(system.file("retentionIndex/metabolites.txt",
```
Comment (jorainer) [inline vignettes/MetaboCoreUtils.Rmd:143]: same here, please change to `system.file("retentionIndex", "metabolites", ...)`.

File: R/retentionIndexing.R [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+#' @export
+#'
+#' @examples
+#'
+#' rti <- data.frame(rtime = c(1,2,3),
+#' rindex = c(100,200,300))
+#' rtime <- c(1.5, 2.5)
+#' rtimeIndexing(rtime, rti)
+#' 
+rtimeIndexing <- function(x,
```
Comment (jorainer) [inline R/retentionIndexing.R:32]: I'm wondering: would it not be better to call this function `rtimeIndex` or `indexRtime` to make it explicit that you want to actually do the indexing. I would call the whole process rtime indexing, but IMHO the function should be called without the *ing*. Thus, to make it clear that you want to *index the retention times* I would favout `indexRtime` over `rtimeIndex` (which would also mean *return the rtime index*)

File: tests/testthat/test_retentionIndexing.R [hunk 7]
```diff
@@ -0,0 +1,17 @@
...
+  
+  rtime <- read.table(system.file("retentionIndex/metabolites.txt",
+                                  package = "MetaboCoreUtils"),
+                      header = TRUE,
+                      sep = "\t")
+  
+  rtime$rindex2 <- rtimeIndexing(rtime$rtime, rti)
+  
+  expect_equal(round(rtime$rindex_manual, 0), round(rtime$rindex2, 0))
+
```
Comment (jorainer) [inline tests/testthat/test_retentionIndexing.R:16]: could you maybe also add some unit tests that check what happens if `x` or `y` are empty, or if `y` does not have the required columns?

**Review by jorainer — CHANGES_REQUESTED** [review]
Nice function @michaelwitting . I have some minor suggestions.

File: inst/retentionIndex/metabolites.txt (hunk 3 above)
Comment (michaelwitting) [reply inst/retentionIndex/metabolites.txt:3]: It is not like an index as you might think of, e.g. from a database. It's more like a conversion to a different "unit". Most people use it as integers, but I would like to keep it as floating point number and let the user decide.

Comment (jorainer) [reply inst/retentionIndex/metabolites.txt:3]: Fine for me - maybe you could describe this also in the documentation?

File: vignettes/MetaboCoreUtils.Rmd (hunk 4 above)
Comment (michaelwitting) [reply vignettes/MetaboCoreUtils.Rmd:138]: Done.

File: vignettes/MetaboCoreUtils.Rmd (hunk 5 above)
Comment (michaelwitting) [reply vignettes/MetaboCoreUtils.Rmd:143]: Done.

File: R/retentionIndexing.R (hunk 6 above)
Comment (michaelwitting) [reply R/retentionIndexing.R:32]: Done.

File: tests/testthat/test_retentionIndexing.R (hunk 7 above)
Comment (michaelwitting) [reply tests/testthat/test_retentionIndexing.R:16]: Done.

**Review by jorainer — APPROVED** [review]
Fine with the PR. you can go ahead an merge.

**Comment by michaelwitting** [comment]
Would love to merge it, but I cannot do it...
